### PR TITLE
Refactor maps.erl

### PR DESCRIPTION
**Pull request by michalmuskala**
* improve performance primarily by using direct pattern matching instead of `maps:find/2`;

* refine types in hopes to improve dialyzer analysis for code using maps.

### CODE REVIEW

File: lib/stdlib/src/maps.erl [hunk 1]
```diff
@@ -231,14 +215,11 @@ filter_1(Pred, Iter) ->
     end.
 
 -spec fold(Fun,Init,MapOrIter) -> Acc when
-    Fun :: fun((K, V, AccIn) -> AccOut),
+    Fun :: fun((Key, Value, AccIn) -> AccOut),
```
Comment (jhogberg) [inline lib/stdlib/src/maps.erl]: The documentation references `K` and `V` through `anno` tags, which need to be updated with the new names.

**Comment by michalmuskala** [comment]
Updated

**Comment by garazdawi** [comment]
The dialyzer test small_SUITE:maps_sum fails with these changes.

**Comment by michalmuskala** [comment]
Should be fixed now.

**Comment by jhogberg** [comment]
Merged, thanks for your contribution!

**Pull request merged**